### PR TITLE
Manage artifacts <DO NOT MERGE>

### PR DESCRIPTION
**Pull request by roguishmountain**
The changes allow the plugin to handle managed artifacts and updates the jenkins version

### CODE REVIEW

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureBlobAction.java [hunk 1]
```diff
@@ -1,34 +1,27 @@
...
-import hudson.model.Run;
-
-import com.microsoftopentechnologies.windowsazurestorage.WAStoragePublisher;
-import com.microsoftopentechnologies.windowsazurestorage.WAStoragePublisher.WAStorageDescriptor;
-import com.microsoftopentechnologies.windowsazurestorage.beans.StorageAccountInfo;
-
 
 public class AzureBlobAction implements RunAction {
-	private final AbstractBuild build;
+	private final Run build;
```
Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureBlobAction.java]: This should actually be removed altogether.  The inclusion of this field causes GC issues, since the Runcannot be unloaded by the GC while the pointer is still live.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureBlobAction.java [hunk 2]
```diff
@@ -64,7 +57,7 @@ public void onBuildComplete() {
 	public void onLoad() {
 	}
 	
-	public AbstractBuild<?,?> getBuild() {
+	public Run<?,?> getBuild() {
```
Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureBlobAction.java]: This can be removed too.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java [hunk 3]
```diff
@@ -159,74 +201,73 @@ public boolean perform(AbstractBuild build, Launcher launcher,
 				}
 			}
 
+			Job<?, ?> job = Jenkins.getInstance().getItemByFullName(projectName, Job.class);
 			// Resolve download location
-			String downloadDir = Utils.replaceTokens(build, listener,
-					downloadDirLoc);
+            BuildFilter filter = new BuildFilter();
```
Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: nit: spacing

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java [hunk 4]
```diff
@@ -159,74 +201,73 @@ public boolean perform(AbstractBuild build, Launcher launcher,
...
+					listener.getLogger()
+							.println(
+									Messages.AzureStorageBuilder_files_downloaded_count(filesDownloaded));
+				}
+			} else if (source instanceof FreeStyleBuild) {
+				downloadArtifacts(source, run, launcher, expIncludePattern, expExcludePattern, listener, strAcc);
+			} else if (source instanceof MatrixBuild) {
+				for (Run r : ((MatrixBuild) source).getExactRuns()) {
+					downloadArtifacts(r, run, launcher, expIncludePattern, expExcludePattern, listener, strAcc);
+				}
```
Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: Do we need special handling for Workflow/Pipeline jobs here?

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStorageClient.java [hunk 5]
```diff
@@ -527,7 +595,7 @@ public static int download(AbstractBuild<?, ?> build,
 			if (Utils.isNullOrEmpty(downloadDirLoc)) {
 				downloadDir = workspacePath;
 			} else {
-				downloadDir = new FilePath(workspacePath, downloadDirLoc);
+                downloadDir = new FilePath(workspacePath, downloadDirLoc);
```
Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStorageClient.java]: nit: spacing

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureBlobAction.java [hunk 6]
```diff
@@ -1,34 +1,27 @@
...
-
 
 public class AzureBlobAction implements RunAction {
-	private final AbstractBuild build;
+	private final Run build;
 	private final String storageAccountName;
 	private final String containerName;
 	private final boolean allowAnonymousAccess;
 	private final AzureBlob zipArchiveBlob;
 	private final List<AzureBlob> individualBlobs;
```
Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureBlobAction.java]: This might be out of scope, but this should probably be kept in a separate file, compressed in some form, and loaded on demand.  The size of this can actually get reallllly large if the number of artifacts archived gets really big.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java [hunk 7]
```diff
@@ -159,74 +201,73 @@ public boolean perform(AbstractBuild build, Launcher launcher,
 				}
 			}
 
+			Job<?, ?> job = Jenkins.getInstance().getItemByFullName(projectName, Job.class);
```
Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: This could probably return null if the job doesn't exist (say, was deleted)

**Comment by mmitche** [comment]
The other thing to do with this (maybe not now) is implement the hooks for job dsl so that configuration can be scripted.

**Comment by directhex** [comment]
> The other thing to do with this (maybe not now) is implement the hooks for job dsl so that configuration can be scripted.

I agree, but I think that ought to be a distinct pull request with a distinct goal

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java (hunk 4 above)
Comment (roguishmountain) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: Yes, they need to be handled differently. It'll be done in a different PR

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java [hunk 8]
```diff
@@ -49,18 +63,29 @@
 	private String downloadDirLoc;
 	private boolean flattenDirectories;
 	private boolean includeArchiveZips;
+	private BuildSelector buildSelector;
+	private String projectName;
 
 	@DataBoundConstructor
-	public AzureStorageBuilder(String storageAccName, String containerName,
+	public AzureStorageBuilder(String storageAccName, String containerName, BuildSelector buildSelector,
```
Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: Does CopyArtifact become a necessary dependency of Azure Storage plugin after this change?

**Comment by mmitche** [comment]
@directhex I can do that change in a separate PR on top of yours.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java [hunk 9]
```diff
@@ -159,74 +201,77 @@ public boolean perform(AbstractBuild build, Launcher launcher,
...
+								.println(
+										Messages.AzureStorageBuilder_files_downloaded_count(filesDownloaded));
+					}
+				} else if (source instanceof FreeStyleBuild) {
+					downloadArtifacts(source, run, launcher, expIncludePattern, expExcludePattern, listener, strAcc);
+				} else if (source instanceof MatrixBuild) {
+					for (Run r : ((MatrixBuild) source).getExactRuns()) {
+						downloadArtifacts(r, run, launcher, expIncludePattern, expExcludePattern, listener, strAcc);
+					}
+				}
```
Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: Any idea whether this interacts properly with pipeline jobs?

**Comment by mmitche** [comment]
LGTM aside from the issue around the Pipeline jobs (they don't derive from Job)

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java (hunk 8 above)
Comment (directhex) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: AFAIK yes - but Sam already added it as a dependency to pom.xml

**Comment by roguishmountain** [comment]
Does the current version support Pipeline jobs?

**Comment by mmitche** [comment]
@snallami PTAL

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureBlobAction.java (hunk 1 above)
Comment (roguishmountain) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureBlobAction.java]: @mmitche This and the getBuild() method need to stay as the side panel jelly uses them to display the azure artifacts

**Comment by mmitche** [comment]
You sure there isn't another way?  That causes memory leaks :/  Maybe a weak reference.

**Comment by roguishmountain** [comment]
AFAIK if the side panel is there, then getBuild() needs to be there

**Comment by mmitche** [comment]
I bet there is a way to get from the side panel back to the build.  I'll do some research (I fixed this bug in a cursory way a long while ago).

Basically what happens is that the builds don't get collected.  A large number of strings gets accumulated over time and eventually Jenkins OOMs.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java [hunk 10]
```diff
@@ -366,5 +422,19 @@ public StorageAccountInfo getStorageAccount(String storageAccountName) {
 			}
 			return storageAcc;
 		}
+
+		public DescriptorExtensionList<BuildSelector, Descriptor<BuildSelector>> getAvailableBuildSelectorList() {
```
Comment (crummel) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureStorageBuilder.java]: Something seems to be going wrong here.  For only a single AzureStorageBuilder, the removed BuildSelectors are not present, but if I add two AzureStorageBuilder steps, all of the BuildSelectors are available and are persisted to the job config if saved.  On reload, the options are still present in the list but the selected option has been reset.

**Comment by roguishmountain** [comment]
@mmitche Have you found a way to get from the side panel back to the build?

**Comment by mmitche** [comment]
@roguishmountain I believe what we ended up doing is storing job name/number (as string/int, so it can be looked up).  This enables collection.  I'm going to put it in live testing this week.  That said other plugins do show stuff on the project page, so I'll also take a look at those to find out how that works

**Comment by crummel** [comment]
@roguishmountain The commit for this is https://github.com/crummel/windows-azure-storage-plugin/commit/70143f648e5fb12826f4548be5022bd6002dd757.  It's definitely a bit messy so I'm still planning to check out other ways to do it.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStorageClient.java [hunk 11]
```diff
@@ -331,6 +323,10 @@ public static int upload(AbstractBuild<?, ?> build, BuildListener listener,
 
 			listener.getLogger().println(
 					Messages.WAStoragePublisher_uploading());
+			
+			if (manageArtifacts) {
+                expContainerName = "jobs";
```
Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStorageClient.java]: Is it possible to avoid hard-coding this?  If we can't have different container names for different jobs (or just custom container names in general) then we should probably at least make this configurable on a global level.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java [hunk 12]
```diff
@@ -265,63 +271,69 @@ public StorageAccountInfo getStorageAccount() {
...
 
 		// Resolve container name
-		String expContainerName = Utils.replaceTokens(build, listener,
-				containerName);
-		if (expContainerName != null) {
+		 String expContainerName = Util.replaceMacro(containerName, envVars);
+        if (!Utils.isNullOrEmpty(expContainerName)) {
+            expContainerName = Utils.FWD_SLASH + expContainerName.trim().toLowerCase(
+                    Locale.ENGLISH) + Utils.FWD_SLASH;
+        } else if (Utils.isNullOrEmpty(expContainerName) && !manageArtifacts) {
```
Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: I don't think the !manageArtifacts condition needs to be here.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java [hunk 13]
```diff
@@ -265,63 +271,69 @@ public StorageAccountInfo getStorageAccount() {
...
 		}
 		if (!Utils.isNullOrEmpty(expVP) && !expVP.endsWith(Utils.FWD_SLASH)) {
 			expVP = expVP.trim() + Utils.FWD_SLASH;
 		}
-
-		// Validate input data
-		if (!validateData(build, listener, strAcc, expContainerName)) {
-			return true; // returning true so that build can continue.
+		
+		if (!Utils.isNullOrEmpty(expVP) && manageArtifacts) {
```
Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: It might be worth prepending the job folder names here.  The job/build number won't be garaunteed to be unique if folders are used in Jenkins

Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/WAStoragePublisher.java]: Actually this is okay.  JOB_NAME is the full path.

File: src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureBlobAction.java [hunk 14]
```diff
@@ -1,42 +1,39 @@
...
 public class AzureBlobAction implements RunAction {
-	private final AbstractBuild build;
 	private final String storageAccountName;
 	private final String containerName;
 	private final boolean allowAnonymousAccess;
 	private final AzureBlob zipArchiveBlob;
 	private final List<AzureBlob> individualBlobs;
+	private final Run build;
 
-	public AzureBlobAction(AbstractBuild build, String storageAccountName, String containerName,
```
Comment (mmitche) [inline src/main/java/com/microsoftopentechnologies/windowsazurestorage/AzureBlobAction.java]: This is typically how this is done without storing a pointer to the BUild:

https://github.com/jenkinsci/build-failure-analyzer-plugin/blob/master/src/main/java/com/sonyericsson/jenkins/plugins/bfa/model/FailureCauseProjectAction.java#L41

You have two actions, one for projects and one for runs.  Runs are just attached normally and contain the data.  Project actions dynamically retrieve the last build and then the action from that build if it exists.  This avoids any back points which cause havoc with the GC.

/cc @crummel

**Comment by patope** [comment]
Stupid question: What how does manage artefacts setting differs, if i set virtual path to value "$JOB_NAME/$BUILD_NUMBER/test-files/virtual" ?

**Comment by roguishmountain** [comment]
@arroyc I made a second PR that has some additional changes and is cleaned up a bit. https://github.com/jenkinsci/windows-azure-storage-plugin/pull/28

**Comment by arroyc** [comment]
Thanks @roguishmountain, I will close this PR then. For now I'm keeping it open to review the comments made ehre, it will help me/my team to review #28 ...
